tsdb: add histogram bucket count to block stats

PaurushGarg opened the pull request:
**Add Native Histogram Bucket Count to Block Stats**
This PR tracks native histogram bucket counts in block statistics.

**What changed:**

- Added `NumHistogramBuckets` to `BlockStats`
- Added `BucketCount()` methods to `Histogram` and `FloatHistogram`
- During compaction, we now iterate over histogram chunks and decode samples to count buckets
- Added histogram cases to `BenchmarkCompaction`

**Performance:**
This introduces some regression since we're now decoding histogram chunks during compaction to count buckets.

Benchmark results:
https://github.com/prometheus/prometheus/pull/17653#issuecomment-3948928356
<!--
If it applies.
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
More at https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-notes block below.
Otherwise, please describe what should be mentioned in the CHANGELOG. Use the following prefixes:
[FEATURE] [ENHANCEMENT] [PERF] [BUGFIX] [SECURITY] [CHANGE]
Refer to the existing CHANGELOG for inspiration:  https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md
If you need help formulating your entries, consult the reviewer(s).
-->
```release-notes
NONE
```

yeya24 commented on tsdb/compact.go:
```diff
@@ -910,8 +930,32 @@ func (DefaultBlockPopulator) PopulateBlock(ctx context.Context, metrics *Compact
+				// Count histogram buckets from all samples
+				iter := chk.Chunk.Iterator(nil)
+				for iter.Next() == chunkenc.ValHistogram {
+					_, h := iter.AtHistogram(nil)
```
We can try to reuse the histogram here?

yeya24 commented on tsdb/compact.go:
```diff
@@ -910,8 +930,32 @@ func (DefaultBlockPopulator) PopulateBlock(ctx context.Context, metrics *Compact
+				// Count float histogram buckets from all samples
+				iter := chk.Chunk.Iterator(nil)
+				for iter.Next() == chunkenc.ValFloatHistogram {
+					_, fh := iter.AtFloatHistogram(nil)
```
ditto

yeya24 commented on tsdb/compact_test.go:
```diff
@@ -1099,9 +1099,22 @@ func TestCompaction_populateBlock(t *testing.T) {
+							if smpl.h.ZeroCount > 0 {
+								bucketCount++
+							}
+							s.NumHistogramBuckets += bucketCount
```
We can just call the helper function here?

yeya24 commented on tsdb/compact.go:
```diff
@@ -778,6 +779,25 @@ func AllSortedPostings(ctx context.Context, reader IndexReader) index.Postings {
+// countFloatHistogramBuckets returns the number of buckets in a float histogram.
+func countFloatHistogramBuckets(fh *histogram.FloatHistogram) uint64 {
+	return countBuckets(len(fh.PositiveBuckets), len(fh.NegativeBuckets), fh.ZeroCount > 0)
+}
```
Is it the right place to have those functions? Should we move these helper functions to histogram package

PaurushGarg commented:
> Hello from the bug-scrub! @PaurushGarg do you think you will come back to this?

@bboreham I have updated the CR to resolve the comments and added benchmark tests and results.

beorn7 commented:
Apologies for the late response. As this touches histograms, I guess I am in charge, at least somewhat. (But note that I'll retire very soon, so I also mention @codesome @jesusvazquez @krajorama here so that this doesn't fall through the cracks.)

Before doing a detailed code review, let's get some high-level concepts clarified:

In my understanding, compaction is generally a very fast procedure because it simply rearranges chunks, without any need to check their content, _unless_ it encounters overlapping chunks (AKA vertical compaction).

With the change here, however, any histogram chunk has to be iterated through, materializing all the histograms in there, just to check their bucket count, so that we can add it to the block stats. (And in case of overlapping chunks, the chunks will be iterated through _twice_, once for the merging, and then again to find out about the bucket count. This could be optimized, but in relative terms, it's less bad because it just doubles the effort, while the non-overlapping case turns from "very cheap" to "very expensive".)

In short, my current gut feeling is that we pay a lot here for one additional value in block stats that is probably not of interest for the overwhelming majority of users.

I see my gut feeling confirmed by looking at the relevant benchmarks, which are the `Compaction/type=histogram...` ones. The worst case represented in benchmarks almost quadruples the compaction time and almost doubles the allocated bytes. I'm not sure how much of a dent that will make in the bigger picture, but given that the feature we add here is probably irrelevant for the vast majority of users, the tolerance for performance regressions is pretty low.

@PaurushGarg @yeya24 for what do you need this? Is it in the Cortex context? Have you run this in production (or at least real-world scenarios)? How much of a performance regression have you seen?
